src/main/scala/cognite/spark/v1/MetricsSource.scala

```diff
@@ -2,22 +2,39 @@ package org.apache.spark.datasource
 
 import cats.Eval
 import com.codahale.metrics._
-import org.apache.spark._
-
 import java.util.concurrent.ConcurrentHashMap
+import java.util.{Timer, TimerTask}
+import org.apache.spark._
+import scala.concurrent.duration._
+import scala.language.postfixOps
 
 class MetricsSource {
   // Add metricNamespace to differentiate with spark system metrics.
   // Keeps track of all the Metric instances that are being published
-  val metricsMap = new ConcurrentHashMap[String, Eval[Counter]]
+  val metricsMap = new ConcurrentHashMap[String, Eval[ExpiringCounter]]
+
+  def getOrCreateCounter(metricNamespace: String, name: String): Counter = {
+    val ctx = Option(TaskContext.get())
+
+    def getNumberOrEmpty(getter: TaskContext => AnyVal): String =
+      ctx.map(getter).map(_.toString()).getOrElse("")
+
+    // The number of fields in the name should be consistent, even if there is no value for
+    // the attempt numbers
+    val stageId = getNumberOrEmpty(_.stageId())
+    val stageAttempt = getNumberOrEmpty(_.stageAttemptNumber())
+    val partitionId = getNumberOrEmpty(_.partitionId())
+    val taskAttempt = getNumberOrEmpty(_.taskAttemptId())
+
+    val metricName = s"$stageId.$stageAttempt.$partitionId.$taskAttempt.$name"
+    val key = s"$metricNamespace.$metricName"
 
-  def getOrCreateCounter(metricNamespace: String, metricName: String): Counter = {
     val wrapped = Eval.later {
       val counter = new Counter
-      registerMetricSource(metricNamespace, metricName, counter)
-      counter
+      val source = registerMetricSource(metricNamespace, metricName, counter)
+      new ExpiringCounter(key, source)
```

Comment:
feels like it may start timer even if entry is already present, then unadded value will unregister prematurely. Could be that Eval.later introduces some laziness that prevents this, but sounds fragile even if it does

Reply:
That is fair, I'd say. The registering of a new metric with the same name can be considered as the metric being "active". Therefore, it will reset the inactivity timer.
Probably fairer than receiving a metric that will die moments later

Or did I misunderstand your concern?

Reply:
If 2 getOrCreateCounter calls create and start 2 counters, only one ends up in the map and receive events, the other will decide nothing happens and will deregister both by metric name